hal/rfcs/0000-spi-overhaul.md

```diff
@@ -0,0 +1,254 @@
+- The supported frequency range must include the range [0.2..2] MHz.
+- The shortest part of the duty cycle must not be shorter than 50% of the expected period.
+- `spi_init()` initializes the pins leaving the configuration registers unchanged.
+- `spi_init()` if `is_slave` is false:
+    - if `ssel` is `NC` the hal implementation ignores this pin.
+    - if `ssel` is not `NC` then the hal implementation owns the pin and its management.
+- When managed by the hal implementation, `ssel` is always considered active low.
+- When the hardware supports the half-duplex (3-wire) mode, if `miso` (exclusive) or `mosi` is missing in any function that expects pins, the bus is assumed to be half-duplex.
+- `spi_free()` resets the pins to their default state.
+- `spi_free()` disables the peripheral clock.
+- `spi_format()` sets :
+  - the number of bits per symbol
+  - the mode :
+    0. Clock idle state is *low*, data are sampled when the clock becomes *active* (polarity = 0, phase = 0)
+    1. Clock idle state is *low*, data are sampled when the clock becomes *inactive* (polarity = 0, phase = 1)
+    2. Clock idle state is *high*, data are sampled when the clock becomes *active* (polarity = 1, phase = 0)
+    3. Clock idle state is *high*, data are sampled when the clock becomes *inactive* (polarity = 1, phase = 1)
+  - the bit ordering (lsb/msb first).
+- `spi_format()` updates the configuration of the peripheral except the baud rate generator.
+- `spi_frequency()` sets the frequency to use during the transfer.
+- `spi_frequency()` returns the actual frequency that will be used.
+- `spi_frequency()` updates the baud rate generator leaving other configurations unchanged.
+- `spi_init()`, `spi_frequency()` and `spi_format()` must be called at least once each before initiating any transfer.
+- `spi_transfer()` :
+  - writes `tx_len` symbols to the bus.
+  - reads `rx_len` symbols from the bus.
+  - if `rx` is NULL then inputs are discarded.
+  - if `tx` is NULL then `fill_symbol` is used instead.
+  - expects symbols types to be the closest stdint type bigger or equal to its size following the platform's endianness.  
+    eg:
+    - 7bits => uint8_t
+    - 15bits => uint16_t
+    - 16bits => uint16_t
+    - 17bits => uint32_t
+  - In Full-duplex mode :
+    - if `rx_len` > `tx_len` then it sends `(rx_len-tx_len)` additional `fill_symbol` to the bus.
+  - In Half-duplex mode :
+    - as master, `spi_transfer()` sends `tx_len` symbols and then reads `rx_len` symbols.
+    - as slave, `spi_transfer()` receives `rx_len` symbols and then sends `tx_len` symbols.
+- `spi_transter_async()` schedules a transfer to be process the same way `spi_transfer()` would have but asynchronously.
+- `spi_transter_async()` returns immediately with a boolean indicating whether the transfer was successfully scheduled or not.
+- The callback given to `spi_transfer_async()` is invoked when the transfer completes (with a success or an error).
```

Comment:
Please consider adding some information regarding `spi_transfer_async()` about parameters like: `void *ctx, DMAUsage hint`.
Information about `event` format passed to async transmission callback is also missing.